File: oj-backend/docker/js/runner.js
const { execFileSync } = require('child_process');
const fs = require('fs');

const DELIMITER = '---TC_END---';

// 1. Decode user code from base64
const code = Buffer.from(process.env.USER_CODE || '', 'base64').toString('utf8');
const wrapper = `
const __input = require('fs').readFileSync(0, 'utf8');
    global.input = __input;
    global.lines = __input.split('\n');
    global.__lineIndex = 0;
    global.readline = () => global.lines[global.__lineIndex++] || '';
`;
fs.writeFileSync('/tmp/solution.js', wrapper + '\n' + code);

// 2. Read all stdin (contains all test case inputs separated by delimiter)
const allInput = fs.readFileSync(0, 'utf8');
const blocks = allInput.split(DELIMITER + '\n').filter(b => b.length > 0);

// 3. Run user code once per test case
for (const block of blocks) {
    try {
        const out = execFileSync('node', ['/tmp/solution.js'], {
            input: block,
            stdio: ['pipe', 'pipe', 'pipe'],
            timeout: 30000, // safety net — real timeout is Docker-level kill
            maxBuffer: 64 * 1024,
        });
        process.stdout.write(out);
    } catch (e) {
        // If the child process errored, forward stderr
        if (e.stderr && e.stderr.length > 0) process.stderr.write(e.stderr);
        if (e.stdout && e.stdout.length > 0) process.stdout.write(e.stdout);
    }
    process.stdout.write('\n' + DELIMITER + '\n');
}

File: oj-backend/docker/python/runner.py
import sys, os, base64, subprocess

DELIMITER = '---TC_END---'

# 1. Decode user code
code = base64.b64decode(os.environ.get('USER_CODE', '')).decode('utf-8')
with open('/tmp/solution.py', 'w') as f:
    f.write(code)

# 2. Read all stdin
all_input = sys.stdin.read()
blocks = [b for b in all_input.split(DELIMITER + '\n') if b.strip()]

# 3. Run user code once per test case
for block in blocks:
    try:
        result = subprocess.run(
            ['python3', '/tmp/solution.py'],
            input=block,
            capture_output=True,
            text=True,
            timeout=30  # safety net
        )
        sys.stdout.write(result.stdout)
        if result.stderr:
            sys.stderr.write(result.stderr)
    except subprocess.TimeoutExpired:
        sys.stderr.write('Time limit exceeded\n')
    except Exception as e:
        sys.stderr.write(str(e) + '\n')
    sys.stdout.write('\n' + DELIMITER + '\n')
    sys.stdout.flush()
